wp-content/plugins/tapestry/controller/class.tapestry-controller.php

```diff
@@ -3,4 +3,110 @@
  * Add/update/retrieve Tapestry post and its child nodes
  * 
  */
-class TapestryController {}
\ No newline at end of file
+
+class TapestryController {
+    const POST_TYPES = [
+        'TAPESTRY' => 'tapestry',
+        'TAPESTRY_NODE' => 'tapestry_node'
+    ];
+    const ERRORS = [
+        'INVALID_POST_ID' => [
+            'MESSAGE' => 'PostID is invalid',
+            'STATUS' => ['status' => 404]
+        ]
+    ];
+
+    /**
+     * Update Tapestry nodes first then
+     * Update the existing Tapestry post if the postId is provided
+     * Otherwise, a new post will be created
+     * 
+     * @param type @post The post data
+     * @param type @postId The postId of the Tapestry
+     */
+    public function updateTapestry($tapestry, $postId = null) {
+        if (!isset($postId))
+            $postId = $this->insertPost($tapestry, 'tapestry');
+        $this->updateNodes($tapestry->nodes, $postId);
+
+        // TODO: Groups and Permisisons data could be added here later
+
+        if (!isset($tapestry->rootId))
+            $tapestry->rootId = $tapestry->nodes[0]->id;
+
+        $tapestry->nodes = $this->getNodeIds($tapestry->nodes);
+        update_post_meta($postId, 'tapestry', $tapestry);
+        return $tapestry;
+    }
+
+    /**
+     * Update Tapestry post's child nodes
+     * 
+     * @param type @nodes An array of child nodes to be updated
+     * @param type @postId The post ID to which the nodes belong
+     * @return WP_Error if postId is invalid
+     */
+    public function updateTapestryNodes($nodes, $postId = null) {
+        if (!isset($postId))
+            return $this->throwsError('INVALID_POST_ID');
+        $this->updateNodes($nodes, $postId);
+        return $nodes;
+    }
+
+    private function throwsError($code) {
+        $ERROR = (object) self::ERRORS[$code];
+        return new WP_Error($code, $ERROR->MESSAGE, $ERROR->STATUS);
+    }
+
+    private function getNodeIds($nodes) {
+        return array_map(function($node) {
+            return $node->id;
+        }, $nodes);
+    }
+
+    private function makeMetadata($node, $nodePostId) {
+        return (object) array(
+            'post_id' => $nodePostId,
+            'title' => $node->title,
+            'coordinates' => (object) array(
+                'x' => $node->fx,
+                'y' => $node->fy
+            )
+        );
+    }
+
+    private function updateNodes($nodes, $postId) {
+        foreach ($nodes as $node) {
+            if (!isset($node->id)) {
+                $nodePostId = $this->insertPost($node, 'tapestry_node');
+                $metadata = $this->makeMetadata($node, $nodePostId);
+                $node->id = add_post_meta($postId, 'tapestry_node', $metadata);
+            } else {
+                $metadata = get_metadata_by_mid('post', $node->id)->meta_value;
+                $nodePostId = $metadata->post_id;
+            }
+            update_post_meta($nodePostId, 'tapestry_node_data', $node);
```

Comment:
Don't you need to add in the tapestry ID somewhere in the meta key so we know this is the node in this tapstry?

Reply:
I don't think we need to add the ID to the meta key anymore. It's because we now have the `post_meta` of each node saved under the **tapestry**, which includes the node's `post_id` pointing to the node details.

In this way, we can enable sharing the node details among different tapestries. Only the node's `post_meta` saved under the **tapestry** is unique to a tapestry. And we use the `post_meta` saved under the **tapestry** to overwrite the node details such as title, coordinates etc. Does it make sense and align with what we discussed?

Reply:
Ok got it. I didn't realize this post meta refers to the node details that are shared.